File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/validation/DateFieldValidation.java
package nerd.tuxmobil.fahrplan.congress.validation;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import nerd.tuxmobil.fahrplan.congress.MyApp;
import nerd.tuxmobil.fahrplan.congress.models.Lecture;
import nerd.tuxmobil.fahrplan.congress.utils.DateHelper;

public class DateFieldValidation {

    private final List<ValidationError> mValidationErrors;

    public DateFieldValidation() {
        mValidationErrors = new ArrayList<>();
    }

    /**
     * Returns a list of validation errors or null.
     */
    public List<ValidationError> getValidationErrors() {
        return mValidationErrors;
    }

    /**
     * Prints all validation errors.
     */
    public void printValidationErrors() {
        for (ValidationError validationError : mValidationErrors) {
            MyApp.LogDebug(getClass().getName(), validationError.toString());
        }
    }

    /**
     * Returns true if the time stamps in the "date" fields of
     * each "event" are within a valid time range.
     * The time range is defined by the "date" attribute of
     * the "day" fields. Otherwise false is returned.
     */
    public boolean validate(@NonNull List<Lecture> lectures) {
        // Order lectures by "dateUtc" field (milliseconds)
        Collections.sort(lectures, (lecture1, lecture2) -> Long.compare(lecture1.getDateUTC(), lecture2.getDateUTC()));
        if (lectures.isEmpty()) {
            return true;
        }

        // Prepare time range (first day)
        Lecture firstLecture = lectures.get(0);
        String firstDateString = firstLecture.date;
        Date firstDate = DateHelper.getDate(firstDateString, "yyyy-MM-dd");
        String formattedFirstDate = DateHelper.getFormattedDate(firstDate);

        // Prepare time range (last day)
        Lecture lastLecture = lectures.get(lectures.size() - 1);
        String lastDateString = lastLecture.date;
        Date lastDate = DateHelper.getDate(lastDateString, "yyyy-MM-dd");
        // Increment date by one day since events also happen on the last day
        // and no time information is given - only the pure date.
        lastDate = DateHelper.shiftByDays(lastDate, 1);
        String formattedLastDate = DateHelper.getFormattedDate(lastDate);

        // Check if the time stamp in <date> is within the time range (first : last day)
        // defined by "date" attribute in the <day> nodes.
        for (Lecture lecture : lectures) {
            validateEvent(lecture, firstDate, lastDate, formattedFirstDate, formattedLastDate);
        }

        // Evaluate validation
        MyApp.LogDebug(getClass().getName(), "Validation result for <date> field: " + mValidationErrors.size() + " errors.");
        return mValidationErrors.isEmpty();
    }

    private void validateEvent(Lecture lecture, Date firstDate, Date lastDate,
                               String formattedFirstDate, String formattedLastDate) {
        long dateUtcInMilliseconds = lecture.dateUTC;
        Date dateUtc = new Date();
        dateUtc.setTime(dateUtcInMilliseconds);

        Date[] dateRange = new Date[]{firstDate, lastDate};
        if (!DateHelper.dateIsWithinRange(dateUtc, dateRange)) {
            String eventId = lecture.lecture_id;
            String formattedDateUtc = DateHelper.getFormattedDate(dateUtc);
            String errorMessage = "Field <date> " + formattedDateUtc + " of event "
                    + eventId +
                    " exceeds range: [ " + formattedFirstDate + " : " + formattedLastDate
                    + " ]";
            ValidationError error = new ValidationError(errorMessage);
            mValidationErrors.add(error);
        }
    }

}

File: network/src/main/java/info/metadude/android/eventfahrplan/network/validation/DateFieldValidation.kt
package info.metadude.android.eventfahrplan.network.validation

import info.metadude.android.eventfahrplan.commons.logging.Logging
import info.metadude.android.eventfahrplan.commons.temporal.DayRange
import info.metadude.android.eventfahrplan.commons.temporal.Moment
import info.metadude.android.eventfahrplan.network.models.Session
import org.threeten.bp.ZoneOffset

internal class DateFieldValidation(

    private val logging: Logging

) {

    private companion object {
        const val LOG_TAG = "DateFieldValidation"
    }

    private val validationErrors: MutableList<ValidationError> = ArrayList()

    fun printValidationErrors() {
        for (validationError in validationErrors) {
            logging.d(LOG_TAG, validationError.toString())
        }
    }

    /**
     * Returns true if the timestamps in the [Session.dateText] fields of each session are within a valid time range.
     * The time range is defined by the [Session.dateText] fields of first and last session (which are sorted by [Session.dateUTC]).
     */
    fun validate(sessions: List<Session>): Boolean {
        val sortedSessions = sessions.sortedBy { it.dateUTC }

        if (sortedSessions.isEmpty()) {
            return true
        }

        val firstDateString = sortedSessions[0].dateText
        val lastDateString = sortedSessions.last().dateText
        val range = DayRange(Moment.parseDate(firstDateString), Moment.parseDate(lastDateString))

        // Check if the time stamp in <date> is within the time range (first : last day)
        // defined by "date" attribute in the <day> nodes.
        sortedSessions.forEach { validateSession(it, range) }

        logging.d(LOG_TAG, "Validation result for <date> field: ${validationErrors.size} errors.")
        return validationErrors.isEmpty()
    }

    private fun validateSession(session: Session, dateRange: DayRange) {
        val dateUtcInMilliseconds = session.dateUTC
        val sessionDate = Moment.ofEpochMilli(dateUtcInMilliseconds).toZonedDateTime(ZoneOffset.UTC)
        if (!dateRange.contains(sessionDate)) {
            val sessionId = session.sessionId
            val errorMessage = ("Field <date> '$sessionDate' of session '$sessionId' exceeds range: [ ${dateRange.startsAt} : ${dateRange.endsAt} ]")
            val error = ValidationError(errorMessage)
            validationErrors.add(error)
        }
    }
}
